README.md
# MNXB01-project
Repo for the final project of the MNXB01 course at Lund University. Autumn 2021.

**Group H**: Ennio Mara, Jesper Larsson and Antton Lamarca

## Usage

The 'rootlogon.C' file makes it so that the main 'project.cpp' file, the 'src/tempTrender.cpp' file and the 'measurement.cpp' file
are compiled automatically with just starting ROOT.
(Make sure to include the local path to your 'include/' directory in the 'rootlogon.C' file for it to work!).

In order to use the implemented functions, you can either run the 'project()' function directly:

    root
    project()

Or create a tempTrender object and call the specific member functions you are interested in with the desired inputs:

    root
    tempTrender* test_2 = new tempTrender("../datasets/smhi-opendata_1_162870_20210926_101011_Lulea.csv")
    test_2->tempOnDay(1,2)

Note that we avoided uploading the data to GitHub, but you will of course need to include the local path to the file of interest in your 'project.cpp' file for ``` project() ``` to work.

## Contents

The structure of the project is simple:

You can find information about the development steps of the project in the 'ChangeLog.txt' file.

All the code is stored in the 'code/' directory. It contains the main 'project.cpp' file as well as the 'rootlogon.C' file.
The 'code/src/' directory contains the 'tempTrender.cpp' file, which has code for three of the proposed example graphs and an original graph that calculates the mean yearly temperature for a city in a given range of years.
The 'measurement.cpp' file contains code for another original graph that calculates average temperatures for the same month over the years.
Header files for both are stored in 'code/include/'.

The 'report/' directory contains the .tex files and images necessary to generate the report. Please run the 'main.tex' file to obtain the final PDF.

## Workplan

Broadly speaking, the separation of tasks was the following:
* Ennio worked on the graph that shows yearly average temperature for a specific month, produced by 'measurement.cpp'.
* Jesper worked on the ``` tempMeanYearly() ``` member function of 'tempTrender.cpp'.
* Antton worked on the non-original member functions of 'tempTrender.cpp': both versions of ``` tempOnDay() ```, and ``` tempPerDay() ```.


ChangeLog.txt
-- okt 25 --
GitHub repo created.

-- okt 28 --
Antton: Started implementing the first of the examples, the histogram of temperatures for a single day.

-- okt 29 --
Antton: Finished rudimentary imlpementation of the first histogram example.

-- okt 31 --
Antton: Finished rudimentary implementation of second example, tempPerDay().

-- nov 01 --
Jesper: Created Function to read data and calculate yearly average temperatures (tempMeanYearly()).

-- nov 02 --
Ennio: Started work of measuring monthly temperature fluctuations, over all years.
Jesper: Created bespoke Function to calculate the average of all vector elements. Bug-fixing on tempMeanYearly().

-- nov 03 --
Antton: Added the tempOnDay() that works with day number. General fixes and tweaks.
Ennio: Finished monthly temperature fluctuations
Jesper: Fixed looping over dates and years in tempMeanYearly(). Added code to graph results.

-- nov 04 --
Antton: Helped fix the tempMeanYearly() function. pushed changes to the PR directly. Added .tex files for report.
Ennio: Changed the monthly temperature fit line to be more clear and saved measurements to png files.
Jesper: Adjusted graph axes for tempMeanYearly(), added grid lines to improve readability.

-- nov 07 --
Ennio: Updated report with the temperature measurements grouped by month.


code/rootlogon.C
void rootlogon() {
  gStyle->SetOptStat(0); //Let's make our plots look a bit better. Get rid of the stat box
  gStyle->SetOptTitle(0); //Get rid of the title (good plots use legends instead)
  gStyle->SetTitleSize(0.05, "x"); //Use bigger text on the axes
  gStyle->SetTitleSize(0.05, "y");
  gStyle->SetLabelSize(0.05, "x"); //Use bigger labels too
  gStyle->SetLabelSize(0.05, "y");
  gStyle->SetPadTopMargin(0.05); //Change the margins to fit our new sizes
  gStyle->SetPadRightMargin(0.05);
  gStyle->SetPadBottomMargin(0.16);
  gStyle->SetPadLeftMargin(0.16);

  // Tell the compiler where to look for headers. Use the same kinds of flags
  //  that you've seen in the course. Note that paths cannot be relative or use
  //  ~
  
  gSystem->AddIncludePath("-I/home/einarelen/project/include -I/home/antton/MNXB01-project/code/include -I/home/je2487la/git/MNXB01-project/code/include -I/Users/enniomara/gitProjects/MNXB01-project/code/include");
  //  You can add multiple flags if you have multiple directories to add
  // gSystem->AddIncludePath("-I/home/einarelen/project/include -I/home/einarelen/project/external/include");
  // Or you can keep adding them one by one, which looks neater
  // gSystem->AddIncludePath("-I/home/einarelen/project/external/include");

  // If you want to add flags to the linking process, use the
  // gSystem->AddLinkedLibs() function. Use the same kinds of flags that you are
  // used to from the course
  //
  // Requires that you have a shared library and that it can be found in the
  // LD_LIBRARY_PATH environment variable
  // gSystem->AddLinkedLibs(" -L ~/project/external/lib64 -lfmt " )
  // If you want to skip that, you can link using the Rpath method
  // gSystem->AddLinkedLibs(" -L/home/einarelen/project/external/lib64 -Wl,-rpath=/home/einarelen/project/external/lib64 -lfmt");

  gROOT->ProcessLine(".L src/measurement.cpp+"); //Load the classes that we want to use automatically - Convenient!
  gROOT->ProcessLine(".L src/tempTrender.cpp+"); //Load the classes that we want to use automatically - Convenient!
  gROOT->ProcessLine(".L project.cpp+"); //The + means to recompile only if it changed sine last time
  //Now you can type project() to invoke the function defined in project
  //You can also create and invoke functions in a temptrender manually by doing e.g. tempTrender t(pathToFile); t.hotCold();
}


code/src/measurement.cpp
#include "measurement.h"
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <string>
#include <vector>

#include "TCanvas.h"
#include "TF1.h"
#include "TGraph.h"
#include "TLegend.h"

Measurement::Measurement(std::string date, double measurement)
    : date(date), measurement(measurement) {}

Measurement measurementFromLine(std::string line) {
  std::stringstream ss(line);
  std::vector<std::string> row;
  std::string cell;

  while (ss.good() && std::getline(ss, cell, ';')) {
    // Initialize and add <colname, int vector> pairs to result
    row.push_back(cell);
  }

  return Measurement(row[0], std::stod(row[2]));
}

std::vector<std::string> Measurement::parse_date() const {
  // why c++ wouln't have date parsing in the std is beyond be
  std::stringstream date(Measurement::date);
  std::string date_item;
  std::vector<std::string> parsed_date_items;
  while (date.good() && getline(date, date_item, '-')) {
    parsed_date_items.push_back(date_item);
  }

  return parsed_date_items;
}
Month Measurement::measurement_month() const {
  auto parsed_date_items = Measurement::parse_date();

  return (Month)std::stoi(parsed_date_items[1]);
}
int Measurement::measurement_year() const {
  auto parsed_date_items = Measurement::parse_date();

  return std::stoi(parsed_date_items[0]);
}

double Measurements::mean() const {
  if (Measurements::measurements.size() == 0) {
    return 0.0;
  }

  double sum{0};
  for (Measurement m : Measurements::measurements) {
    sum += m.measurement;
  }

  return sum / Measurements::measurements.size();
}
void Measurements::push_back(Measurement m) {
  Measurements::measurements.push_back(m);
}

void run(std::string path_to_file, int month_int) {
  Month month = (Month)month_int;
  std::ifstream file(path_to_file);

  std::string line;
  // Read until we encouter "Datum;", which means that the next lines will be
  // the actual data
  while (getline(file, line)) {
    if (line.rfind("Datum;", 0) == 0) {
      break;
    }
  }

  Measurements measurements{};
  while (getline(file, line)) {
    Measurement measurement = measurementFromLine(line);
    if (measurement.measurement_month() == month) {
      measurements.push_back(measurementFromLine(line));
    }
  }

  // bucket measurements per year
  std::map<int, Measurements> measurements_per_year;
  for (Measurement m : measurements.measurements) {
    measurements_per_year[m.measurement_year()].push_back(m);
  }

  std::cout << measurements.measurements.size() << std::endl;
  std::cout << measurements.measurements[0].measurement_month() << std::endl;
  std::cout << measurements_per_year.size() << std::endl;

  TCanvas *c4 = new TCanvas("c4", "4 - Example yearly mean graph", 200, 10, 500, 300);
  TGraph *gr = new TGraph();

  for (auto p : measurements_per_year) {
    gr->SetPoint(gr->GetN(), p.first, p.second.mean());
  }

  TF1 *f = new TF1("func", "[1] + [0]*x", 1800, 2025);
  f->SetLineStyle(1);
  f->SetLineWidth(6);
  f->SetLineColor(8);
  gr->Fit(f);
  f->Draw();

  // A - draw axis
  // L - draw line between points
  // * - draw points
  gr->Draw("AL*");

  std::string out = "measurement" + std::to_string(month_int) + ".png";
  c4->SaveAs(out.c_str());
}


code/include/measurement.h
#ifndef MEASUREMENT_H
#define MEASUREMENT_H

#include <string>
#include <vector>
enum Month {
  jan = 1,
  feb = 2,
  mar = 3,
  apr = 4,
  may = 5,
  jun = 6,
  jul = 7,
  aug = 8,
  sep = 9,
  oct = 10,
  nov = 11,
  dec = 12
};

class Measurement {
public:
  Measurement() = default;
  Measurement(std::string date, double measurement);

  Month measurement_month() const;
  int measurement_year() const;
  const std::string date = "";
  const double measurement = 0;

private:
  std::vector<std::string> parse_date() const;
};

class Measurements {
public:
  double mean() const;
  void push_back(Measurement);

  std::vector<Measurement> measurements;
};

// takes a comma delimetered line from the dataset, and builds a measurement object
Measurement measurementFromLine(std::string line);

void run(std::string, int);

#endif


code/include/tempTrender.h
#ifndef TEMPTRENDER_H
#define TEMPTRENDER_H

#include <string>
class tempTrender {
	public:
	explicit tempTrender(const std::string& filePath); //Construct using the specified file
	~tempTrender() {} //Destructor

	//Member functions
	void tempOnDay(int monthToCalculate, int dayToCalculate) const; //Make a histogram of the temperature on this day
	void tempOnDay(int dateToCalculate) const; //Make a histogram of the temperature on this date
	void tempPerDay() const; //Make a histogram of the average temperature of each day of the year
	// void hotCold() const; //Make a histogram of the hottest and coldest day of the year
	// void tempPerYear(int yearToExtrapolate) const; //Make a histogram of average temperature per year, then fit and extrapolate to the given year
	void tempMeanYearly(int yearStart, int yearEnd) const;
	//Member variables
    std::string _path;

	private:


};
#endif


code/src/tempTrender.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <TGraph.h>
#include<TH1F.h>
#include<TProfile.h>
#include<TLegend.h>
#include <TCanvas.h> // canvas object
#include "tempTrender.h"

using namespace std;

//Utils
bool hasEnding (std::string const &fullString, std::string const &ending) {
  //compares two strings to check if the first string ends with the second string
  //e.g. hasEnding("binary", "nary") returns True
  //Function taken from https://stackoverflow.com/questions/874134/find-out-if-string-ends-with-another-string-in-c
  //maybe should go in its own file and then get included here ?
	if (fullString.length() >= ending.length()) {
	    return (0 == fullString.compare (fullString.length() - ending.length(), ending.length(), ending));
	} else {
	    return false;
	}
}

int date_to_number (int month, int day) {
  //takes month and date and returns the corresponding # out of 365 days in a year
  int month_sum = 0; //contribution of months to the day number
  int days_in_month [12] = { 0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30};
  for (int i=0; i<month; i++) {
    month_sum += days_in_month[i];
  }
  return month_sum+day;
}

double VecAvg(vector<double> v){ //Average of vector elements
  double sum = 0;
  int denom = v.size();
  for(int i=0; i < denom; i++){
    sum += v[i];
  }
  return sum/denom;
}

//Member Functions
// #0
tempTrender::tempTrender(const string& filePath) : _path{filePath}{ //filePath has now been stored in member variable _path
  cout << "The user supplied " << filePath <<" as the path to the data file.\n";
  //std::cout << "You should probably store this information in a member variable of the class! Good luck with the project! :)\n";

}
// #1
void tempTrender::tempOnDay(int monthToCalculate, int dayToCalculate) const { //Make a histogram of the temperature on this day

  //Incorrect input handling
  if (monthToCalculate>12 || dayToCalculate>31) {
    cout << "Invalid date requested. Please input a valid date with the format tempOnDay(Month, Day)" << endl;
    return;
  }

  //Convert inputed ints to strings, and combine them in "02-14" format
  string month_prefix = "";
  if (monthToCalculate < 10) { 
    month_prefix = "0";
  }
  string day_prefix = "";
  if (dayToCalculate < 10) {
    day_prefix = "0";
  }
  string dateToCalculate_string = month_prefix+to_string(monthToCalculate)+"-"+day_prefix+to_string(dayToCalculate);
  cout<<"The requested date was " << dateToCalculate_string << endl;

  //Code to open the csv files
  fstream fin; //File pointer
  fin.open(_path, ios::in); //Open file at '_path'
  if (fin.fail()) { //Check that file can be opened.
    cout << "File could not be opened. Please check that the provided path is correct." << endl;
	return;
  }

  // CREATE PROFILE HISTOGRAM that we will fill with data from specified date
	//(name, title;xlabel;ylabel, bins, xmin, xmax)
  TH1D* hDayTemp = new TH1D("one_day_tempt", "Temperature for date;Temperature [\xb0 C]; Entries", 100, -20, 40); 

  //Iterate through file, line by line, checking if the date matches with input
  vector<string> row;
  string line, cell, date_string;
  int temp;
  int i = 0;
  while (getline(fin, line)){ //read whole file, row by row, store line in variable 'line' each loop
    i++;
    row.clear();

    stringstream lineStream(line); //Slice line by ; and store each part in vector 'row'
	while (lineStream.good() && getline(lineStream, cell, ';')) {
        row.push_back(cell);
    }
	date_string = row[0]; //save the date of the line
    //If the date matches with requested date
    if (hasEnding(date_string, dateToCalculate_string)) {
      temp = stoi(row[2]);
      //cout << "line " << i << ", at date " << date_string << " the temperature was " << temp << endl;
      hDayTemp->Fill(temp);
    }

  }

  //Histogram drawing
  TCanvas* c1 = new TCanvas("c1", "1 - histogram canvas", 900, 600);
  hDayTemp->SetFillColorAlpha(kRed, 0.35);
  hDayTemp->Draw();

  //Legend
  //TODO: make legend nicer
  auto legend = new TLegend();
  legend->AddEntry(hDayTemp,("Temperature on "+dateToCalculate_string).c_str(),"f");
  legend->Draw();

  //Answering questions
  Double_t mean_val = hDayTemp->GetMean(1);
  Double_t stdev = hDayTemp->GetRMS(1);
  cout << "The mean value for date " << dateToCalculate_string << " is: " << mean_val << "C" << endl;
  cout << "The standard deviation is: " << stdev << endl;

  //TODO: Fit gaussian to histogram to calulcate probabilities?
}

// #2
void tempTrender::tempOnDay(int dateToCalculate) const { //Make a histogram of the temperature on this date

  //Incorrect input handling
  if (dateToCalculate>365) {
    cout << "Invalid date requested. Please input a valid day number between 1 and 365." << endl;
    return;
  }
  cout<<"The requested day was day number " << dateToCalculate << endl;

  //Code to open the csv files
  fstream fin; //File pointer
  fin.open(_path, ios::in); //Open file at '_path'
  if (fin.fail()) { //Check that file can be opened.
    cout << "File could not be opened. Please check that the provided path is correct." << endl;
	return;
  }

  // CREATE PROFILE HISTOGRAM that we will fill with data from specified date
	//(name, title;xlabel;ylabel, bins, xmin, xmax)
  TH1D* hDayTemp2 = new TH1D("one_day_tempt_2", "Temperature for date;Temperature [\xb0 C]; Entries", 100, -20, 40);

  //Iterate through file, line by line, checking if the date matches with input
  vector<string> row, year_month_day;
  string line, cell, date_string, date_item;
  int month, day, temp;
  int i = 0;
  while (getline(fin, line)){ //read whole file, row by row, store line in variable 'line' each loop
    i++;
    if(i>13) { //Skip header lines of the csv file. Not the cleanest but will do for now.
	
      row.clear();
      year_month_day.clear();

      stringstream lineStream(line); //Slice line by ; and store each part in vector 'row'
	  while (lineStream.good() && getline(lineStream, cell, ';')) {
          row.push_back(cell);
      }
      temp = stoi(row[2]); //save the temperature in the lane
	  date_string = row[0]; //save the date of the line
      stringstream dateStream(date_string); //split 'date_string' into year, month, day using 'getline()', save results in 'year_mont_day' vector
      while (dateStream.good() && getline(dateStream, date_item, '-')) { 
        year_month_day.push_back(date_item);
      }
      month = stoi(year_month_day[1]);
      day = stoi(year_month_day[2]);
      if (month==2 && day==29) { //Deal with leap years by skipping 02/29
        continue;
      }
      //If the date matches with requested date
      if (date_to_number(month,day)==dateToCalculate) {
        temp = stoi(row[2]);
        //cout << "line " << i << ", at date " << date_string << " the temperature was " << temp << endl;
        hDayTemp2->Fill(temp);
      }
    }
  }

  //Histogram drawing
  TCanvas* c2 = new TCanvas("c2", "2 - histogram canvas", 900, 600);
  hDayTemp2->SetFillColorAlpha(kRed, 0.35);
  hDayTemp2->Draw();

  //Legend
  //TODO: make legend nicer
  auto legend2 = new TLegend();
  legend2->AddEntry(hDayTemp2,("Temperature on day "+to_string(dateToCalculate)).c_str(), "f");
  legend2->Draw();

}

// #3
void tempTrender::tempPerDay() const { //Make a histogram of the average temperature of each day of the year

  //Code to open the csv files
  fstream fin; //File pointer
  fin.open(_path, ios::in); //Open file at '_path'
  if (fin.fail()) { //Check that file can be opened.
    cout << "File could not be opened. Please check that the provided path is correct." << endl;
	return;
  }

  // CREATE TProfile HISTOGRAM, specifying option "s" to draw standard deviation
	//(name, title;xlabel;ylabel, bins, xmin, xmax, option)
  auto* hTempPerDay = new TProfile("temp_per_day", "Temperature per day;Day of year;Temperature [\xb0 C]", 365, 1, 365, "s"); 

  //Iterate through file, line by line, checking if the date matches with input
  vector<string> row, year_month_day;
  string line, cell, date_string, date_item;
  int month, day, temp;
  while (getline(fin, line)){ //read whole file, row by row, store line in variable 'line' each loop
    row.clear();
    year_month_day.clear();
    stringstream lineStream(line); //Slice line by ; and store each part in vector 'row'
	while (lineStream.good() && getline(lineStream, cell, ';')) {
        row.push_back(cell);
    }
	date_string = row[0]; //save the date of the line
    if (any_of(date_string.begin(), date_string.end(), ::isdigit)) { //crude check to see if 'date_string' is indeed a date
      temp = stoi(row[2]);

      stringstream dateStream(date_string); //split 'date_string' into year, month, day using 'getline()', save results in 'year_mont_day' vector
      while (dateStream.good() && getline(dateStream, date_item, '-')) { 
        year_month_day.push_back(date_item);
      }
      month = stoi(year_month_day[1]);
      day = stoi(year_month_day[2]);
      if (month==2 && day==29) { //Deal with leap years by skipping 02/29
        continue;
      }
      hTempPerDay->Fill(date_to_number(month,day), temp); //Fill temp value in corresponding bin
    } 
  }
  TCanvas* c3 = new TCanvas("c3", "3 - TProfile histogram canvas", 900, 600);
  hTempPerDay->Draw();
}
// void tempTrender::hotCold() const {} //Make a histogram of the hottest and coldest day of the year
// void tempTrender::tempPerYear(int yearToExtrapolate) const {} //Make a histogram of average temperature per year, then fit and extrapolate to the given year

// #4
void tempTrender::tempMeanYearly(int yearStart, int yearEnd) const { 
  //create a line-graph showin mean-yearly temperature over time.
  
  cout<<"The requested time period was " << yearStart << "-" << yearEnd << endl;

  //Create graph to store/display our data
  auto graph1 = new TGraph();
  //Code to open the csv files
  fstream fin; //File pointer
  fin.open(_path, ios::in); //Open file at '_path'
  if (fin.fail()) { //Check that file can be opened.
    cout << "File could not be opened. Please check that the provided path is correct." << endl;
	  return;
  };
  //Iterate through file, line by line, checking if the date matches with input
  vector<string> row, rowdate;
  
  string line, cell, date_string, YearString, Last_Date;
  double tempentry; 
  double tempdailysum = 0; 
  double tempyearlysum = 0;
  int sumentries = 0;
  vector<double> tempdailyaverage, YearlyAverage;
  double  tempyearlyaverage;
  int i = 0 ;
  int k= 0;
  int Year_Last = 0;
  int Year_Count = 0;
  int YearCurrent = 0;
  bool RepeatedDate = false;

  while (getline(fin, line)){ //read whole file, row by row, store line in variable 'line' each loop
    i++;
    row.clear();

    stringstream lineStream(line); //Slice line by ; and store each part in vector 'row'
	while (lineStream.good() && getline(lineStream, cell, ';')) {
      row.push_back(cell);
    }
	if(i >= 20){ //Does this mean that the user can't select the first year of a dataset?
              // Yes. Too bad!
      date_string = row[0]; //save the date of the line
      if (Last_Date==date_string) {
        RepeatedDate = true;
      } else {
        RepeatedDate = false;
      }
      YearString = date_string.substr(0, 4);
      YearCurrent = stoi(YearString);// Save year to integer
    } else {
      continue;
    }

  
  //If the year is within the specified range
    if (YearCurrent>=yearStart && YearCurrent<= yearEnd) {
      tempentry = stoi(row[2]); //Save temp as int

      //Check wheter current date matches last
      if (RepeatedDate){ //If yes: add temp entry to the sum of the days entries, increase sum of the entries by one
        tempdailysum += tempentry ;
        sumentries++ ;
      } else { //New day we haven't seen before
        if (Last_Date.empty()) { //first loop
          tempdailysum += tempentry ;
          sumentries++ ;
        } else { //all other loops beyond the first
          tempdailyaverage.push_back((tempdailysum / sumentries)); //if not: add sum to vector containing all daily averages 
          tempdailysum = 0; //Reset tempdailysum & sumentries
          sumentries = 0;

          //Start calculation for new date
          tempdailysum += tempentry;
          sumentries++;

          if(Year_Last == 0 || Year_Last!=YearCurrent){ //If year has changed, sum up all daily entries and average them out.
            Year_Count++; //keep count of years we've seen
            //cout << Year_Last << YearCurrent << "Happy New Year" << endl;
            tempyearlyaverage = VecAvg(tempdailyaverage);
            tempdailyaverage.clear();
            YearlyAverage.push_back(tempyearlyaverage);
            graph1->SetPoint(k,YearCurrent, tempyearlyaverage);
            k++ ;
          }
        }
      }
    
    } else { //if year not in the desired range
      continue;
    }
    //update a copy of the date to check against in the next iteration
    Last_Date = date_string;
    Year_Last = YearCurrent;

  }

  cout << "The average temperatures of " << YearlyAverage.size() << " years were calculated" << endl;
  TCanvas* c5 = new TCanvas("c5", "5- Yearly average temperature", 900, 600);
	c5->DrawFrame(yearStart-2,5,yearEnd+2,11.5);
	gPad->SetGrid(1, 1);
  graph1->Draw();
}
